File: src/leetcode/binary_search/875_koko_eating_bananas.ts
const minEatingSpeed = (piles: number[], hrs: number): number => {
  let minHrs = 1;
  let maxHrs = Math.max(...piles);

  while (minHrs < maxHrs) {
    const mid = Math.floor(minHrs + (maxHrs - minHrs) / 2);

    const hoursSpent = piles.reduce((sum, pile) => sum + Math.ceil(pile / mid), 0);

    if (hoursSpent <= hrs) {
      maxHrs = mid;
    } else {
      minHrs = mid + 1;
    }
  }

  return minHrs;
};

export default minEatingSpeed;

File: src/leetcode/binary_search/875_koko_eating_bananas.rb
# frozen_string_literal: true

# @param {Integer[]} piles
# @param {Integer} hrs
# @return {Integer}
def min_eating_speed(piles, hrs)
  min_hrs = 1
  max_hrs = piles.max

  while min_hrs < max_hrs
    mid = min_hrs + ((max_hrs - min_hrs) / 2)

    hours_spent = piles.sum { |pile| (pile / mid.to_f).ceil }

    if hours_spent <= hrs
      max_hrs = mid
    else
      min_hrs = mid + 1
    end
  end

  min_hrs # at the end of while, min == max
end
